### PR TITLE
Add spec for connection state freshness check

### PR DESCRIPTION
**Pull request by SimonWoolf**
Cf https://github.com/ably/realtime/issues/940#issuecomment-286985784
Prioritised due to analysis in https://github.com/ably/realtime/issues/1289#issuecomment-341152556
Fixes https://github.com/ably/docs/issues/287
Ably-js implementation https://github.com/ably/ably-js/blob/c55b4518d6f096e4bde90afcd04c5ba35d466f3c/common/lib/transport/connectionmanager.js#L702-L712

### CODE REVIEW

**Comment by paddybyers** [comment]
> Ably-js implementation https://github.com/ably/ably-js/blob/c55b4518d6f096e4bde90afcd04c5ba35d466f3c/common/lib/transport/connectionmanager.js#L702-L712

But here: https://github.com/ably/ably-js/blob/c55b4518d6f096e4bde90afcd04c5ba35d466f3c/common/lib/transport/connectionmanager.js#L709

You unconditionally go into `suspended` without knowing that the `suspended` time has passed?

**Comment by SimonWoolf** [comment]
> You unconditionally go into suspended without knowing that the suspended time has passed?
Merge state

If I haven't had any evidence of websocket liveness from realtime for longer than connectionStateTtl + maxIdleInterval, then I do know that the suspended time has passed (assuming that realtime hasn't heard from me for the same duration I haven't heard from realtime -- which isn't certain as there can be different latencies in both directions, but it's a decent approximation). True, I haven't had 2 minutes of timers elapse, but the whole point of RTN15g1 is that I can't rely on 2 minutes of timers having elapsed if the device has been sleeping

**Comment by SimonWoolf** [comment]
(And it doesn't make sense to clear the connection state and not have the connecting failstate be `suspended` -- the `suspended` state is for when we've lost connection state)

**Comment by paddybyers** [comment]
> longer than connectionStateTtl + maxIdleInterval

Yes, sorry, I missed the `connectionStateTtl` in there ...

File: content/client-lib-development-guide/features.textile [hunk 1]
```diff
@@ -360,7 +360,11 @@ h3(#realtime-connection). Connection
...
+*** @(RTN15g1)@ This check should be made before each connection attempt. It is generally not sufficient to merely clear the connection state when moving to @SUSPENDED@ state (though that may be done too), since the device may have been sleeping / suspended, in which case it may have been many hours since it was last actually connected, even though, having been in the @CONNECTED@ state when it was put to sleep, it has only moved out of that state very recently (after waking up and noticing it's no longer connected)
+*** @(RTN15g2)@ Anothing consequence of that is that the measure of whether the client been disconnected for too long (for the purpose of this check) cannot just be whether the client left the @CONNECTED@ state more than @connectionStateTtl@ ago. Instead, it should be whether the difference between the current time and the last activity time is greater than the sum of the @connectionStateTtl@ and the @maxIdleInterval@, where the last activity time is the time of the last known actual sign of activity from Ably per "RTN23a":#RTN23a
```
Comment (mattheworiordan) [inline content/client-lib-development-guide/features.textile]: @SimonWoolf typo, `Anothing`?

Comment (SimonWoolf) [reply content/client-lib-development-guide/features.textile]: Oops, fixed

**Pull request merged**